File: cs50x/week1/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

//For some reason, I tried to do this purely with modulos and without using arrays.
int main(void)
{
    // Ask user for their card number (kinda sketchy)
    long card_number = get_long("Card Number: ");
    // printf("You gave me %li\n", card_number);

    // Checksum - Introduce our variables
    int sum = 0;
    int length = 0;
    //Save the number to get the first two digits later, once we have the length.
    long first_two = card_number;

    //Add every last digit, every other one being multiplied by two.
    while (card_number > 0)
    {
        if (length % 2 == 0)
        {
            sum += card_number % 10;
        }
        //If last digit is 5 (leading to a double digit), break it into two digits.
        else
        {
            if (card_number % 10 > 4)
            {
                sum += 1 + ((card_number % 10 * 2) % 10);
            }
            else
            {
                sum += card_number % 10 * 2;
            }
        }
        //Remove the last digit and add one to length.
        card_number -= card_number % 10;
        card_number /= 10;
        length += 1;
        // printf("Card number: %li\n", card_number);
        // printf("Length: %i\n", length);
    }
    // printf("You got %i.\n", sum);
    for (int x = length; x > 2; x--)
    {
        first_two -= first_two % 10;
        first_two /= 10;
    }
    // printf("First two digits is %li.\n", first_two);
    if (sum % 10 == 0 && (12 < length))
    {
        if (length == 15)
        {
            if (first_two == 34 || first_two == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (length == 13)
        {
            if ((first_two - first_two % 10) / 10 == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (length == 16)
        {
            if (first_two > 50 && first_two < 56)
            {
                printf("MASTERCARD\n");
            }
            else if ((first_two - first_two % 10) / 10 == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

File: cs50x/week6/pset6/credit/credit.py
from cs50 import get_string
import sys
cardnum = get_string("Number: ")

checksum = 0
others = 0

# Way to go through with only one loop, adding every other digit to every other sum.
for i in range(len(cardnum) - 1, -1, -1):
    if len(cardnum) % 2 == i % 2:
        if (int(cardnum[i]) * 2) < 10:
            checksum += (int(cardnum[i]) * 2)
        else:
            checksum += 1 + ((int(cardnum[i]) * 2) % 10)
    else:
        others += int(cardnum[i])

# If elif tree.
if (checksum + others) % 10 == 0:
    if (len(cardnum) == 15):
        if ((cardnum[0] == "3") and (cardnum[1] == "4" or cardnum[1] == "7")):
            print("AMEX")
    elif (len(cardnum) == 16):
        if (cardnum[0] == "5") and (int(cardnum[1]) > 0 and int(cardnum[1]) < 6):
            print("MASTERCARD")
        elif (cardnum[0] == "4"):
            print("VISA")
    elif (len(cardnum) == 13):
        if (cardnum[0] == "4"):
            print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")